x/staking/migrations/v3/store.go

```diff
@@ -0,0 +1,146 @@
+package v3
+
+import (
+	"sort"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"github.com/cosmos/cosmos-sdk/x/staking/exported"
+	"github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+// subspace contains the method needed for migrations of the
+// legacy Params subspace
+type subspace interface {
+	GetParamSet(ctx sdk.Context, ps paramtypes.ParamSet)
+	HasKeyTable() bool
+	WithKeyTable(paramtypes.KeyTable) paramtypes.Subspace
+	Set(ctx sdk.Context, key []byte, value interface{})
+}
+
+// keeper contains the staking keeper functions required
+// for the migration
+type keeper interface {
+	GetAllDelegations(ctx sdk.Context) []types.Delegation
+	GetAllValidators(ctx sdk.Context) []types.Validator
+	SetDelegation(ctx sdk.Context, delegation types.Delegation)
+	SetValidator(ctx sdk.Context, validator types.Validator)
+	RefreshTotalLiquidStaked(ctx sdk.Context) error
+}
+
+// Adds the following LSM params:
+// - ValidatorBondFactor
+// - GlobalLiquidStakingCap
+// - ValidatorLiquidStakingCap
+func MigrateParamsStore(ctx sdk.Context, paramstore subspace) {
+	if !paramstore.HasKeyTable() {
+		paramstore.WithKeyTable(types.ParamKeyTable())
+	}
+	paramstore.Set(ctx, types.KeyValidatorBondFactor, types.DefaultValidatorBondFactor)
+	paramstore.Set(ctx, types.KeyGlobalLiquidStakingCap, types.DefaultGlobalLiquidStakingCap)
+	paramstore.Set(ctx, types.KeyValidatorLiquidStakingCap, types.DefaultValidatorLiquidStakingCap)
+}
+
+// Set each validator's ValidatorBondShares and LiquidShares to 0
+func MigrateValidators(ctx sdk.Context, k keeper) {
+	for _, validator := range k.GetAllValidators(ctx) {
+		validator.ValidatorBondShares = sdk.ZeroDec()
+		validator.LiquidShares = sdk.ZeroDec()
+		k.SetValidator(ctx, validator)
+	}
+}
+
+// Set each delegation's ValidatorBond field to false
+func MigrateDelegations(ctx sdk.Context, k keeper) {
+	for _, delegation := range k.GetAllDelegations(ctx) {
+		delegation.ValidatorBond = false
+		k.SetDelegation(ctx, delegation)
+	}
+}
+
+// migrateUBDEntries will remove the ubdEntries with same creation_height
+// and create a new ubdEntry with updated balance and initial_balance
+func migrateUBDEntries(ctx sdk.Context, store storetypes.KVStore, cdc codec.BinaryCodec, legacySubspace exported.Subspace) error {
```

Comment:
Can we have a test for this?

I'm not sure it exists on cosmos-sdk v47?

It can be setup like so:
1. create old entries -> have multiple entries at the same block height
2. migrate entries -> confirm that entries at the same height got merged

I'm also willing to accept this if this function is tested somewhere, but I don't see that being the case.

Reply:
unit test added in latest commits